From 6ed384111aaac3944d46ccba097ac5310fe34833 Mon Sep 17 00:00:00 2001
From: Bartosz Bilas <b.bilas@grinn-global.com>
Date: Thu, 2 Oct 2025 13:08:24 +0200
Subject: [PATCH] plat: mediatek: use libdram to pass actual memory size to
 U-Boot

Since libdram can obtain the actual memory size at runtime, make it possible
to pass this value to U-Boot. This allows the bootloader to adapt to boards
with different DRAM sizes without requiring manual configuration.
---
 plat/mediatek/mt8188/bl2_plat_setup.c       | 4 ++++
 plat/mediatek/mt8188/include/libdram.h      | 1 +
 plat/mediatek/mt8188/include/platform_def.h | 2 ++
 3 files changed, 7 insertions(+)

diff --git a/plat/mediatek/mt8188/bl2_plat_setup.c b/plat/mediatek/mt8188/bl2_plat_setup.c
--- a/plat/mediatek/mt8188/bl2_plat_setup.c
+++ b/plat/mediatek/mt8188/bl2_plat_setup.c
@@ -398,6 +398,10 @@ void bl2_platform_setup(void)
 	load_partition_table(GPT_IMAGE_ID);
 	blkdev_set_dramk_data_offset(get_part_addr("dramk"));
 	mt_mem_init();
+
+	BOOT_ARGUMENT->dram_size = platform_memory_size();
+	BOOT_ARGUMENT->magic_number = BOOT_ARGUMENT_MAGIC;
+
 	/* change emmc read buffer to DRAM */
 	boot_dev_spec->buffer.offset = 0x41000000;
 	boot_dev_spec->buffer.length = 0x1000000;
diff --git a/plat/mediatek/mt8188/include/libdram.h b/plat/mediatek/mt8188/include/libdram.h
--- a/plat/mediatek/mt8188/include/libdram.h
+++ b/plat/mediatek/mt8188/include/libdram.h
@@ -14,5 +14,6 @@
 void i2c_hw_init(void);
 unsigned int pmic_init(void);
 void mt_mem_init(void);
+unsigned long long platform_memory_size(void);
 
 #endif /* LIBDRAM_H */
diff --git a/plat/mediatek/mt8188/include/platform_def.h b/plat/mediatek/mt8188/include/platform_def.h
--- a/plat/mediatek/mt8188/include/platform_def.h
+++ b/plat/mediatek/mt8188/include/platform_def.h
@@ -303,6 +303,8 @@
 #define BL2_BASE		(0x201000)
 #define BL2_LIMIT		(0x400000)
 
+#define BOOT_ARGUMENT_LOCATION	(0x40000100)
+
 #define MAX_IO_DEVICES			U(3)
 #define MAX_IO_HANDLES			U(4)
 #define MAX_IO_BLOCK_DEVICES			1
-- 
2.48.1
